kv/kvserver: TestStoreRangeGossipOnSplits failed

kv/kvserver.TestStoreRangeGossipOnSplits [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8380955?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8380955?buildTab=artifacts#/) on master @ [0775fcc3b3a71b47dc67a73f2caadbf4b586ec37](https://github.com/cockroachdb/cockroach/commits/0775fcc3b3a71b47dc67a73f2caadbf4b586ec37):


```
=== RUN   TestStoreRangeGossipOnSplits
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/7334302ae8c22802de8334e77d6901c9/logTestStoreRangeGossipOnSplits4002577389
    test_log_scope.go:79: use -show-logs to present logs inline
    client_split_test.go:2391: gossiped range count 55 more than 1 away from expected 81
    client_split_test.go:2397: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/7334302ae8c22802de8334e77d6901c9/logTestStoreRangeGossipOnSplits4002577389
--- FAIL: TestStoreRangeGossipOnSplits (0.38s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeGossipOnSplits.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-23591

kv/kvserver.TestStoreRangeGossipOnSplits [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/8387486?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/8387486?buildTab=artifacts#/) on master @ [d698e5eaf36bee19901beeff7db3425315402907](https://github.com/cockroachdb/cockroach/commits/d698e5eaf36bee19901beeff7db3425315402907):


```
=== RUN   TestStoreRangeGossipOnSplits
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/7438ab096787a15a8ce5f502b5724798/logTestStoreRangeGossipOnSplits2746565145
    test_log_scope.go:79: use -show-logs to present logs inline
    client_split_test.go:2391: gossiped range count 55 more than 1 away from expected 81
    client_split_test.go:2397: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/7438ab096787a15a8ce5f502b5724798/logTestStoreRangeGossipOnSplits2746565145
--- FAIL: TestStoreRangeGossipOnSplits (0.66s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeGossipOnSplits.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStoreRangeGossipOnSplits [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8382823?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8382823?buildTab=artifacts#/) on master @ [b38da1d1c5da65cd6855e302296c578de05f0ab2](https://github.com/cockroachdb/cockroach/commits/b38da1d1c5da65cd6855e302296c578de05f0ab2):


```
=== RUN   TestStoreRangeGossipOnSplits
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreRangeGossipOnSplits351835565
    test_log_scope.go:79: use -show-logs to present logs inline
    client_split_test.go:2391: gossiped range count 55 more than 1 away from expected 81
    client_split_test.go:2397: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreRangeGossipOnSplits351835565
--- FAIL: TestStoreRangeGossipOnSplits (0.26s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeGossipOnSplits.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/client_split_test.go
+++ b/pkg/kv/kvserver/client_split_test.go
@@ -2308,6 +2308,12 @@ func TestStoreRangeGossipOnSplits(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
+	// 50% for testing We can't properly test how frequently changes in the
+	// number of ranges trigger the store to gossip its capacities if we have
+	// to worry about changes in the number of leases also triggering store
+	// gossip.
+	overrideCapacityFraction := 0.5
+
 	ctx := context.Background()
 	serv, _, _ := serverutils.StartServer(t, base.TestServerArgs{
 		Knobs: base.TestingKnobs{
@@ -2316,11 +2322,8 @@ func TestStoreRangeGossipOnSplits(t *testing.T) {
 				DisableSplitQueue: true,
 				DisableScanner:    true,
 				GossipTestingKnobs: kvserver.StoreGossipTestingKnobs{
-					OverrideGossipWhenCapacityDeltaExceedsFraction: 0.5, // 50% for testing
-					// We can't properly test how frequently changes in the number of ranges
-					// trigger the store to gossip its capacities if we have to worry about
-					// changes in the number of leases also triggering store gossip.
-					DisableLeaseCapacityGossip: true,
+					OverrideGossipWhenCapacityDeltaExceedsFraction: overrideCapacityFraction,
+					DisableLeaseCapacityGossip:                     true,
 				},
 			},
 		},
@@ -2385,7 +2388,10 @@ func TestStoreRangeGossipOnSplits(t *testing.T) {
 		}
 		select {
 		case rangeCount = <-rangeCountCh:
-			changeCount := int32(math.Ceil(math.Max(float64(lastRangeCount)*0.5, 10)))
+			changeCount := int32(math.Ceil(math.Max(
+				float64(lastRangeCount)*overrideCapacityFraction,
+				kvserver.GossipWhenRangeCountDeltaExceeds,
+			)))
 			diff := rangeCount - (lastRangeCount + changeCount)
 			if diff < -1 || diff > 1 {
 				t.Errorf("gossiped range count %d more than 1 away from expected %d", rangeCount, lastRangeCount+changeCount)

--- a/pkg/kv/kvserver/store_gossip.go
+++ b/pkg/kv/kvserver/store_gossip.go
@@ -44,11 +44,11 @@ const (
 	//  gossip interval.
 	gossipWhenLeaseCountDeltaExceeds = 5
 
-	// gossipWhenRangeCountDeltaExceeds specifies the absolute change from the
+	// GossipWhenRangeCountDeltaExceeds specifies the absolute change from the
 	// last gossiped store capacity range count which needs to be exceeded
 	// before the store will gossip immediately without waiting for the
 	// periodic gossip interval.
-	gossipWhenRangeCountDeltaExceeds = 5
+	GossipWhenRangeCountDeltaExceeds = 5
 
 	// gossipWhenLoadDeltaExceedsFraction specifies the fraction from the last
 	// gossiped store capacity load which needs to be exceeded before the store
@@ -394,9 +394,6 @@ const (
 // immediate gossip of this store's descriptor, to include updated
 // capacity information.
 func (s *StoreGossip) MaybeGossipOnCapacityChange(ctx context.Context, cce CapacityChangeEvent) {
-	if s.knobs.DisableLeaseCapacityGossip && (cce == LeaseAddEvent || cce == LeaseRemoveEvent) {
-		return
-	}
 
 	// Incrementally adjust stats to keep them up to date even if the
 	// capacity is gossiped, but isn't due yet to be recomputed from scratch.
@@ -460,15 +457,17 @@ func (s *StoreGossip) shouldGossipOnCapacityDelta() (should bool, reason string)
 	updateForWPS, deltaWPS := deltaExceedsThreshold(
 		s.cachedCapacity.lastGossiped.WritesPerSecond, s.cachedCapacity.cached.WritesPerSecond,
 		gossipMinAbsoluteDelta, gossipWhenLoadDeltaExceedsFraction)
-
 	updateForRangeCount, deltaRangeCount := deltaExceedsThreshold(
 		float64(s.cachedCapacity.lastGossiped.RangeCount), float64(s.cachedCapacity.cached.RangeCount),
-		gossipWhenRangeCountDeltaExceeds, gossipWhenCapacityDeltaExceedsFraction)
-
+		GossipWhenRangeCountDeltaExceeds, gossipWhenCapacityDeltaExceedsFraction)
 	updateForLeaseCount, deltaLeaseCount := deltaExceedsThreshold(
 		float64(s.cachedCapacity.lastGossiped.LeaseCount), float64(s.cachedCapacity.cached.LeaseCount),
 		gossipWhenLeaseCountDeltaExceeds, gossipWhenCapacityDeltaExceedsFraction)
 	s.cachedCapacity.Unlock()
+
+	if s.knobs.DisableLeaseCapacityGossip {
+		updateForLeaseCount = false
+	}
 
 	if updateForQPS {
 		reason += fmt.Sprintf("queries-per-second(%.1f) ", deltaQPS)

